pkg/operator/configobserver/apiserver/observe_audit.go

```diff
@@ -0,0 +1,88 @@
+
+	// GetDefaultAuditPolicyPath returns the path to the default audit policy.
+	GetDefaultAuditPolicyPath() string
+}
+
+// NewAuditObserver returns an ObserveConfigFunc that observes the audit field of the APIServer resource
+// and sets the apiServerArguments:audit-policy-file field for the apiserver appropriately.
+func NewAuditObserver(pathGetter AuditPolicyPathGetter) configobserver.ObserveConfigFunc {
+	var (
+		apiServerArgumentsAuditPath = []string{"apiServerArguments", "audit-policy-file"}
+	)
+
+	return func(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+		errs := []error{}
+
+		// does the current config already have an audit policy path set? if so we should make an effort to
+		// return it as the observed config in case of error, otherwise we resort to the default policy path.
+		auditPolicyPath := pathGetter.GetDefaultAuditPolicyPath()
+		current, _, err := unstructured.NestedStringSlice(existingConfig, apiServerArgumentsAuditPath...)
+		if err != nil {
+			// best-effort approach, let's keep going.
+			errs = append(errs, err)
+		}
+		if len(current) > 0 {
+			auditPolicyPath = current[0]
+		}
+
+		defaultObservedConfig := map[string]interface{}{}
+		if err := unstructured.SetNestedStringSlice(defaultObservedConfig, []string{auditPolicyPath}, apiServerArgumentsAuditPath...); err != nil {
+			// If we are here, then we are not setting any default for observed audit profile.
+			// looks like returning the existing config is the best alternative here. This puts a constraint on the
+			// apiserver to have a default configuration in its bindata.
+			return existingConfig, append(errs, err)
+		}
+
+		listers := genericListers.(APIServerLister)
+		apiServer, err := listers.APIServerLister().Get("cluster")
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				klog.Warningf("apiserver.config.openshift.io/cluster: not found")
+
+				// return the audit policy path from the current config if set, otherwise we use the default configuration.
+				return defaultObservedConfig, errs
+			}
+
+			return defaultObservedConfig, append(errs, err)
+		}
+
+		desiredProfile := string(apiServer.Spec.Audit.Profile)
+		if len(desiredProfile) == 0 {
+			return defaultObservedConfig, errs
+		}
+
+		desiredAuditPolicyPath, err := pathGetter.GetAuditPolicyPath(desiredProfile)
+		if err != nil {
+			return defaultObservedConfig, append(errs, fmt.Errorf("audit profile is not valid name=%s", desiredProfile))
+		}
+		if desiredAuditPolicyPath == auditPolicyPath {
+			return defaultObservedConfig, errs
+		}
+
+		observedConfig := map[string]interface{}{}
+		if err := unstructured.SetNestedStringSlice(observedConfig, []string{desiredAuditPolicyPath}, apiServerArgumentsAuditPath...); err != nil {
+			return defaultObservedConfig, append(errs, fmt.Errorf("failed to set desired audit profile in observed config name=%s", desiredProfile))
```

Comment:
return old config, not default. Avoids flapping.

Reply:
it's returning the current config, I changed the variable name. I think it boils down to the question I posted here - https://github.com/openshift/library-go/pull/845#discussion_r459668601.